### PR TITLE
[SYCL] Add template parameter support for num_simd_work_items attribute

### PR DESCRIPTION
**Pull request by smanna12**
This patch

1. adds support for template parameter on [[intelfpga::num_simd_work_items()]] attribute

2. splits test/SemaSYCL/num_simd_work_items.cpp to separate files for
   host compilation (test/SemaSYCL/num_simd_work_items_host.cpp) and
   device compilation (test/SemaSYCL/num_simd_work_items_device.cpp)

3. makes changes to intel_reqd_sub_group_size attribute source codes
   to avoid duplication and reuse for the template parameter support.

Signed-off-by: Soumi Manna <soumi.manna@intel.com>

### CODE REVIEW

File: clang/include/clang/Sema/Sema.h [hunk 1]
```diff
@@ -12836,6 +12834,31 @@ class Sema final {
   }
 };
 
+template <typename AttrType>
+void Sema::addIntelSYCLFunctionAttr(Decl *D, const AttributeCommonInfo &Attr,
```
Comment (MrSidims) [inline clang/include/clang/Sema/Sema.h:12838]:
```suggestion
void Sema::addIntelSYCLSingleArgFunctionAttr(Decl *D, const AttributeCommonInfo &Attr,
```

There are more (triple arguments) function attributes, so I suggest this change.

Comment (smanna12) [reply clang/include/clang/Sema/Sema.h]: Done

File: clang/include/clang/Sema/Sema.h [hunk 2]
```diff
@@ -12836,6 +12834,32 @@ class Sema final {
   }
 };
 
+template <typename AttrType>
+void Sema::addIntelSYCLSingleArgFunctionAttr(Decl *D,
+                                             const AttributeCommonInfo &CI,
+                                             Expr *E) {
+  if (!E)
+    return;
```
Comment (erichkeane) [inline clang/include/clang/Sema/Sema.h:12842]: Should this be an assert?  When is this allowable?

I see that it should have already asserted if !E, so an assert seems more reasonable here.

**Review by erichkeane — COMMENTED** [review]
1 nit, otherwise i both like the direction and implementation of this patch.

**Review by MrSidims — DISMISSED** [review]
LGTM
Thanks!

**Review by premanandrao — APPROVED** [review]
I was good with your changes, until I saw Erich's nit.  :-)
Please address that, but approved.

File: clang/include/clang/Sema/Sema.h (hunk 2 above)
Comment (smanna12) [reply clang/include/clang/Sema/Sema.h]: Yes, i have added assert here.

Comment (smanna12) [reply clang/include/clang/Sema/Sema.h]: @Erich, several codegen tests are failing like this after adding assertion here:

 /export/iusers/smanna/my_sycl_ws/llvm/clang/include/clang/Sema/Sema.h:12841: void clang::Sema::addIntelSYCLSingleArgFunctionAttr(clang::Decl*, const clang::AttributeCommonInfo&, clang::Expr*) [with AttrType = clang::IntelReqdSubGroupSizeAttr]: Assertion `!E' failed.

Comment (elizabethandrews) [reply clang/include/clang/Sema/Sema.h]: Assert should be that E exists. Right now you're asserting !E

Comment (elizabethandrews) [reply clang/include/clang/Sema/Sema.h]: Actually no. I was confused. Disregard my previous comment

File: clang/include/clang/Sema/Sema.h [hunk 3]
```diff
@@ -12836,6 +12834,31 @@ class Sema final {
   }
 };
 
+template <typename AttrType>
+void Sema::addIntelSYCLSingleArgFunctionAttr(Decl *D,
+                                             const AttributeCommonInfo &CI,
+                                             Expr *E) {
+  assert(!E);
```
Comment (erichkeane) [inline clang/include/clang/Sema/Sema.h:12841]: Assert is the wrong way, also needs a message.

File: clang/include/clang/Sema/Sema.h (hunk 2 above)
Comment (erichkeane) [reply clang/include/clang/Sema/Sema.h]: Disregard @elizabethandrews LAST message, her first one was right :)

Comment (elizabethandrews) [reply clang/include/clang/Sema/Sema.h]: Lol yes. I need a break :D

File: clang/include/clang/Sema/Sema.h (hunk 3 above)
Comment (smanna12) [reply clang/include/clang/Sema/Sema.h]: Sorry about this. Thanks Erich. Fixed now.

**Review by erichkeane — APPROVED** [review]

File: clang/include/clang/Sema/Sema.h (hunk 2 above)
Comment (smanna12) [reply clang/include/clang/Sema/Sema.h]: > Assert should be that E exists. Right now you're asserting !E
Yes, i was in wrong way. Sorry about this. Thanks Elizabeth.

**Review by premanandrao — APPROVED** [review]

**Comment by smanna12** [comment]
Thank you, everyone for the reviews.

**Pull request merged**